_posts/2019-01-21-api-best-practices.md

```diff
@@ -0,0 +1,90 @@
+rolling rate limit windows)  
+
+And while we're discussing rate limiting, you may also want to consider returning the appropriate error codes
+when those rate limits are exceeded.  **Error 429 (Too Many Requests)** is the official standard here.
+
+### Use proper status codes
+Pay close attention to what your API is doing, and what sorts of error cases it will encounter.  For almost all
+of the common errors that you might encounter, there's an HTTP status [*that maps directly to that particular
+type of error*](https://tools.ietf.org/html/rfc7231#section-6).  Use it!
+
+There's more to this than just being pedantic, by the way.  Since most developers will be interacting with HTTP
+through some kind of library (like Ruby's [RestClient](https://github.com/rest-client/rest-client), for instance),
+there's a good chance that the library provides support for fine grained error handling based on the specific error
+code being returned.  If you're not fully leveraging the HTTP status codes, then you're depriving your end users of
+that functionality.
+
+### Return structured error messages
+Speaking of errors - a status code is all well and good, if it's the right status code, but additional error messages
+are highly useful.  The HTTP standard allows you to return a body along with any error code, so be helpful and put
+additional information there. To really do it right, structure your error messages so that they can potentially
+contain more than just a brief message - a JSON encoded hash will work perfectly.  You can associate the error
+message with the same key every time, and include additional information as needed, like so:
+
+```
+{"msg":"User quota exceeded","max_allowed":10000,"administrator_id":1234}
+```
+
+In this example you not only get a message that tells you what went wrong (you're over quota), you're also told
+what your quota actually is, and the user id of the person to go bother about it.  Much more useful than just a
+status code or a generic error message alone.
+
+The gold standard for an API that does this well is the [GitHub API](https://developer.github.com/v3/).  Not only
+do you get a structured error message with a description, you also get *a link to the relevant documentation*.
+
+Which brings us to the next thing every good API needs:
+
+### Documentation, Documentation, Documentation
+
+All of the thoughtful design in the world won't mean a thing if no one knows how to use it.  Document every facet
+of your API - the individual endpoints, what they require, what is optional, and what they return, including errors
+and headers.  It's almost impossible to overdo this part.  If you find yourself wondering wether you should document
+some aspect of your API, the answer is "Yes."  Writing documentation is rarely fun, and writing it well is difficult.
+The difficulty is worth it, though, because the strength of documentation alone can often be a deciding factor in wether
+an API actually picks up users.
```

Comment:
@rdormer I don't know if this fits in this category, but libraries like Swagger make this easy: https://medium.com/@sushildamdhere/how-to-document-rest-apis-with-swagger-and-ruby-on-rails-ae4e13177f5d

Reply:
@etagwerker I've used Swagger with APIs in the past, and it was kind of a mixed bag.  So the question is, if I point to Swagger, is that an endorsement of it?  I don't know if I want to endorse it per se.  On the other hand it could just be for informational purposes.  Your thoughts there?

Reply:
@rdormer I think it is good for informational purposes, you don't need to endorse it. You can be upfront about your experience using Swagger and let the reader decide.

Reply:
@etagwerker done.